### PR TITLE
staging/vhost-device-sound: add testing module for sound

### PR DESCRIPTION
**Pull request by dorindabassey**
### Summary of the PR

This PR adds testing modules for the following files in order to satisfy functional testing and code coverage.

```
Filename                       Regions    Missed Regions     Cover   Functions  Missed Functions  Executed       Lines      Missed Lines     Cover    Branches   Missed Branches     Cover
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
audio_backends.rs                   12                 8    33.33%           6                 5    16.67%          25                17    32.00%           0                 0         -
audio_backends/alsa.rs             279               279     0.00%          14                14     0.00%         374               374     0.00%           0                 0         -
audio_backends/null.rs              15                 2    86.67%           7                 0   100.00%          30                 0   100.00%           0                 0         -
audio_backends/pipewire.rs         193                89    53.89%          26                 1    96.15%         468               131    72.01%           0                 0         -
device.rs                          388               262    32.47%          50                19    62.00%         835               417    50.06%           0                 0         -
lib.rs                             103                62    39.81%          32                13    59.38%         188                86    54.26%           0                 0         -
main.rs                             15                 5    66.67%          10                 4    60.00%          32                10    68.75%           0                 0         -
stream.rs                           90                20    77.78%          33                 9    72.73%         278                22    92.09%           0                 0         -
virtio_sound.rs                     67                62     7.46%          56                51     8.93%          56                51     8.93%           0                 0         -
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
TOTAL                             1162               789    32.10%         234               116    50.43%        2286              1108    51.53%           0                 0         -

```

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: staging/vhost-device-sound/src/audio_backends/pipewire.rs [hunk 1]
```diff
@@ -463,3 +464,147 @@ impl AudioBackend for PwBackend {
...
+            descriptor: Descriptor::new(next_addr, 0x200, VRING_DESC_F_NEXT as u16, index + 1),
+            vring,
+        }
+    }
+
+    #[test]
+    fn test_pipewire_backend_success() {
+        let streams = Arc::new(RwLock::new(vec![Stream {
+            ..Stream::default()
+        }]));
```
Comment (epilys) [inline staging/vhost-device-sound/src/audio_backends/pipewire.rs:549]:
```suggestion
        let streams = Arc::new(RwLock::new(vec![Stream::default()]));
```

File: staging/vhost-device-sound/src/audio_backends/pipewire.rs [hunk 2]
```diff
@@ -463,3 +464,147 @@ impl AudioBackend for PwBackend {
...
+    #[should_panic(expected = "Stream does not exist")]
+    fn test_pipewire_backend_panics() {
+        let stream_params = Arc::new(RwLock::new(vec![]));
+
+        let pw_backend = PwBackend::new(stream_params);
+
+        let msg = ctrlmsg();
+        let _ = pw_backend.set_parameters(0, msg);
+        let _ = pw_backend.prepare(0);
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/audio_backends/pipewire.rs:580]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/audio_backends/pipewire.rs [hunk 3]
```diff
@@ -463,3 +464,147 @@ impl AudioBackend for PwBackend {
...
+    #[test]
+    #[should_panic]
+    fn test_pipewire_prepare_panics() {
+        let stream_params = Arc::new(RwLock::new(vec![]));
+        let pw_backend = PwBackend::new(stream_params);
+        let _ = pw_backend.prepare(0);
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/audio_backends/pipewire.rs:587]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/audio_backends/pipewire.rs [hunk 4]
```diff
@@ -463,3 +464,147 @@ impl AudioBackend for PwBackend {
...
+    #[test]
+    #[should_panic]
+    fn test_pipewire_start_panics() {
+        let stream_params = Arc::new(RwLock::new(vec![]));
+        let pw_backend = PwBackend::new(stream_params);
+        let _ = pw_backend.start(0);
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/audio_backends/pipewire.rs:594]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/audio_backends/pipewire.rs [hunk 5]
```diff
@@ -463,3 +464,147 @@ impl AudioBackend for PwBackend {
...
+    #[test]
+    #[should_panic]
+    fn test_pipewire_stop_panics() {
+        let stream_params = Arc::new(RwLock::new(vec![]));
+        let pw_backend = PwBackend::new(stream_params);
+        let _ = pw_backend.stop(0);
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/audio_backends/pipewire.rs:601]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/device.rs [hunk 6]
```diff
@@ -54,15 +54,16 @@ impl VhostUserSoundThread {
     ) -> Result<Self> {
         queue_indexes.sort();
 
-        Ok(Self {
+        let thread = VhostUserSoundThread {
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:57]:
```suggestion
        let thread = Self {
```

Clippy would complain about this (not using `Self`)

File: staging/vhost-device-sound/src/device.rs [hunk 7]
```diff
@@ -54,15 +54,16 @@ impl VhostUserSoundThread {
...
             event_idx: false,
             mem: None,
             chmaps,
             jacks,
             queue_indexes,
             streams,
             streams_no,
-        })
+        };
+        Ok(thread)
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:66]: Wait, this change does not do anything does it?

File: staging/vhost-device-sound/src/device.rs [hunk 8]
```diff
@@ -128,10 +129,14 @@ impl VhostUserSoundThread {
         vring: &VringRwLock,
         audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
     ) -> IoResult<bool> {
+        let atomic_mem = match &self.mem {
+            Some(m) => m,
+            None => return Err(Error::NoMemoryConfigured.into()),
+        };
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:135]:
```suggestion
        let Some(ref atomic_mem) = self.mem else {
            return Err(Error::NoMemoryConfigured.into()),
        };
```

File: staging/vhost-device-sound/src/device.rs [hunk 9]
```diff
@@ -450,10 +455,14 @@ impl VhostUserSoundThread {
...
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:461]:
```suggestion
        let Some(ref atomic_mem) = self.mem else {
            return Err(Error::NoMemoryConfigured.into()),
        };
```

File: staging/vhost-device-sound/src/device.rs [hunk 10]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+
+        let config = SoundConfig::new(SOCKET_PATH.to_string(), false, BACKEND);
+
+        let chmaps = Arc::new(RwLock::new(vec![]));
+        let jacks = Arc::new(RwLock::new(vec![]));
+        let queue_indexes = vec![1, 2, 3];
+        let streams = Arc::new(RwLock::new(vec![Stream {
+            ..Stream::default()
+        }]));
+        let streams_no = 3;
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:847]: Is it 3?

File: staging/vhost-device-sound/src/device.rs [hunk 11]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+            .handle_event(CONTROL_QUEUE_IDX, &vrings, &audio_backend)
+            .is_ok());
+
+        let vring = VringRwLock::new(mem, 0x1000).unwrap();
+        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+        vring.set_queue_ready(true);
+        assert!(t.process_control(&vring, &audio_backend).is_ok());
+        assert!(t.process_tx(&vring, &audio_backend).is_ok());
+        assert!(t.process_rx(&vring, &audio_backend).is_ok());
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:876]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/device.rs [hunk 12]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+
+        let audio_backend =
+            RwLock::new(alloc_audio_backend(config.audio_backend, streams.clone()).unwrap());
+
+        let vring = VringRwLock::new(mem, 0x1000).unwrap();
+        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+        vring.set_queue_ready(true);
+        assert!(t.process_control(&vring, &audio_backend).is_err());
+        assert!(t.process_tx(&vring, &audio_backend).is_err());
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:905]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/device.rs [hunk 13]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+    #[test]
+    fn test_sound_backend() {
+        const SOCKET_PATH: &str = "vsound.socket";
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:908]: I think you should hoist this definition for the entire `tests` module since it's used more than once.

File: staging/vhost-device-sound/src/device.rs [hunk 14]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+        const BACKEND: BackendType = BackendType::Null;
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:909]: And I think it would be better if there were no `const BACKEND` definitions and the value was inlined, like:

```rust
let config = SoundConfig::new(SOCKET_PATH.to_string(), 
false,
BackendType::Null);
```

File: staging/vhost-device-sound/src/device.rs [hunk 15]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+
+        let config = SoundConfig::new(SOCKET_PATH.to_string(), false, BACKEND);
+        let backend = VhostUserSoundBackend::new(config);
+
+        assert!(backend.is_ok());
+        let mut backend = backend.unwrap();
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:915]:
```suggestion
        let mut backend = VhostUserSoundBackend::new(config).expect("Could not create backend.");
```

File: staging/vhost-device-sound/src/device.rs [hunk 16]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+        let ret = backend.handle_event(TX_QUEUE_IDX, EventSet::IN, &vrings, 0);
+        assert!(ret.is_ok());
+        assert!(!ret.unwrap());
+
+        let ret = backend.handle_event(RX_QUEUE_IDX, EventSet::IN, &vrings, 0);
+        assert!(ret.is_ok());
+        assert!(!ret.unwrap());
+
+        // cleanup
+        let _ = std::fs::remove_file(SOCKET_PATH);
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:979]: Since tests will be running in parallel unless --test-threads=1 is set, you should use a temporary directory. See https://docs.rs/tempfile/latest/tempfile/fn.tempdir.html

File: staging/vhost-device-sound/src/device.rs [hunk 17]
```diff
@@ -814,3 +823,204 @@ impl Drop for ControlMessage {
...
+    }
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:980]:
```suggestion
    }

```

File: staging/vhost-device-sound/src/device.rs (hunk 16 above)
Comment (epilys) [reply staging/vhost-device-sound/src/device.rs:979]: Is this why serial_test is used for? If there's no other reason to run serialized tests other than path conflicts, it should be removed.

File: staging/vhost-device-sound/src/lib.rs [hunk 18]
```diff
@@ -105,7 +107,7 @@ pub enum BackendType {
     Alsa,
 }
 
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
```
Comment (epilys) [inline staging/vhost-device-sound/src/lib.rs:110]: Maybe go all the way since we're changing this line:

```suggestion
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

**Review by epilys — CHANGES_REQUESTED** [review]
THe direction looks overall pretty good to me. Needs some minor changes. We should check if `serial_test` is actually needed like I mention in one comment. Also, many lines that setup VRings and other low-level interfaces could use a bunch of comments to explain what's going on.

File: staging/vhost-device-sound/src/device.rs (hunk 6 above)
Comment (dorindabassey) [reply staging/vhost-device-sound/src/device.rs:57]: I think that was a mistake, I don't remember why I changed that. returning Ok(self{}) is fine.

File: staging/vhost-device-sound/src/device.rs [hunk 19]
```diff
@@ -814,3 +820,213 @@ impl Drop for ControlMessage {
...
+        assert!(thread.is_ok());
+        let mut t = thread.unwrap();
+
+        // Artificial memory
+        let mem = GuestMemoryAtomic::new(
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap(),
+        );
+        t.mem = Some(mem.clone());
+
+        // Artificial Vring for queues
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:855]:
```suggestion
        // Mock Vring for queues
```

File: staging/vhost-device-sound/src/device.rs [hunk 20]
```diff
@@ -814,3 +820,213 @@ impl Drop for ControlMessage {
...
+        let config = SoundConfig::new(socket_path, false, BackendType::Null);
+        let mut backend = VhostUserSoundBackend::new(config).expect("Could not create backend.");
+
+        assert_eq!(backend.num_queues(), NUM_QUEUES as usize);
+        assert_eq!(backend.max_queue_size(), 64);
+        assert_ne!(backend.features(), 0);
+        assert!(!backend.protocol_features().is_empty());
+        backend.set_event_idx(false);
+
+        // Artificial memory
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:915]:
```suggestion
        // Mock memory
```

File: staging/vhost-device-sound/src/device.rs [hunk 21]
```diff
@@ -814,3 +820,213 @@ impl Drop for ControlMessage {
...
+        let mem = GuestMemoryAtomic::new(
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap(),
+        );
+
+        // Artificial Vring for queues
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:920]:
```suggestion
        // Mock Vring for queues
```

File: staging/vhost-device-sound/src/device.rs [hunk 22]
```diff
@@ -814,3 +820,213 @@ impl Drop for ControlMessage {
...
+            .path()
+            .join("sound_failures.socket")
+            .display()
+            .to_string();
+        let config = SoundConfig::new(socket_path, false, BackendType::Null);
+        let backend = VhostUserSoundBackend::new(config);
+
+        let mut backend = backend.unwrap();
+
+        // Artificial memory
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:990]:
```suggestion
        // Mock memory
```

File: staging/vhost-device-sound/src/device.rs [hunk 23]
```diff
@@ -814,3 +820,213 @@ impl Drop for ControlMessage {
...
+        let mem = GuestMemoryAtomic::new(
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap(),
+        );
+
+        // Artificial Vring for queues
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:995]:
```suggestion
        // Mock Vring for queues
```

File: staging/vhost-device-sound/src/stream.rs [hunk 24]
```diff
@@ -282,3 +282,223 @@ impl Drop for Buffer {
...
+            status: VIRTIO_SND_S_OK.into(),
+            desc_chain: prepare_desc_chain::<VirtioSndPcmSetParams>(GuestAddress(0), hdr, 1),
+            descriptor: Descriptor::new(next_addr, 0x200, VRING_DESC_F_NEXT as u16, 1),
+        }
+    }
+
+    #[test]
+    fn test_pcm_state_transitions() {
+        let mut state = PCMState::new();
+        assert_eq!(state.to_string(), "VIRTIO_SND_R_PCM_SET_PARAMS");
```
Comment (epilys) [inline staging/vhost-device-sound/src/stream.rs:361]:
```suggestion
        assert_eq!(state, PCMState::SetParameters);
```

File: staging/vhost-device-sound/src/stream.rs [hunk 25]
```diff
@@ -282,3 +282,223 @@ impl Drop for Buffer {
...
+
+        assert!(state.set_parameters().is_ok());
+        state.set_parameters().unwrap();
+        assert_eq!(state.to_string(), "VIRTIO_SND_R_PCM_SET_PARAMS");
```
Comment (epilys) [inline staging/vhost-device-sound/src/stream.rs:365]:
```suggestion
        assert_eq!(state, PCMState::SetParameters);
```

File: staging/vhost-device-sound/src/stream.rs [hunk 26]
```diff
@@ -282,3 +282,223 @@ impl Drop for Buffer {
...
+
+        assert!(state.prepare().is_ok());
+        state.prepare().unwrap();
+        assert_eq!(state.to_string(), "VIRTIO_SND_R_PCM_PREPARE");
```
Comment (epilys) [inline staging/vhost-device-sound/src/stream.rs:369]:
```suggestion
        assert_eq!(state, PCMState::Prepare);
```

File: staging/vhost-device-sound/src/stream.rs [hunk 27]
```diff
@@ -282,3 +282,223 @@ impl Drop for Buffer {
...
+
+        state.release().unwrap();
+        assert_eq!(state.to_string(), "VIRTIO_SND_R_PCM_RELEASE");
```
Comment (epilys) [inline staging/vhost-device-sound/src/stream.rs:372]:
```suggestion
        assert_eq!(state, PCMState::Release);
```

File: staging/vhost-device-sound/src/stream.rs [hunk 28]
```diff
@@ -282,3 +282,223 @@ impl Drop for Buffer {
...
+    }
+
+    #[test]
+    fn test_invalid_state_transition() {
+        let mut state = PCMState::new();
+        assert_eq!(state.to_string(), "VIRTIO_SND_R_PCM_SET_PARAMS");
```
Comment (epilys) [inline staging/vhost-device-sound/src/stream.rs:378]:
```suggestion
        assert_eq!(state, PCMState::SetParameters);
```

**Review by epilys — CHANGES_REQUESTED** [review]
Some minor changes. Also, reminder to run `cargo fmt` or `cargo +nightly fmt` until we get it added in the CI checks. I ran it and many lines were affected, see diff:

<details>

<summary>diff</summary>

```diff
diff --git a/staging/vhost-device-sound/src/audio_backends/pipewire.rs b/staging/vhost-device-sound/src/audio_backends/pipewire.rs
index 6583bc9..ffb18d3 100644
--- a/staging/vhost-device-sound/src/audio_backends/pipewire.rs
+++ b/staging/vhost-device-sound/src/audio_backends/pipewire.rs
@@ -467,9 +467,6 @@ impl AudioBackend for PwBackend {
 
 #[cfg(test)]
 mod tests {
-    use super::*;
-    use crate::ControlMessageKind;
-    use crate::SoundDescriptorChain;
     use vhost_user_backend::{VringRwLock, VringT};
     use virtio_bindings::bindings::virtio_ring::{VRING_DESC_F_NEXT, VRING_DESC_F_WRITE};
     use virtio_queue::{mock::MockSplitQueue, Descriptor, Queue, QueueOwnedT};
@@ -477,6 +474,9 @@ mod tests {
         Address, ByteValued, GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap,
     };
 
+    use super::*;
+    use crate::{ControlMessageKind, SoundDescriptorChain};
+
     // Prepares a single chain of descriptors for request queue
     fn prepare_desc_chain<R: ByteValued>(
         start_addr: GuestAddress,
diff --git a/staging/vhost-device-sound/src/device.rs b/staging/vhost-device-sound/src/device.rs
index 6a8d464..0c51844 100644
--- a/staging/vhost-device-sound/src/device.rs
+++ b/staging/vhost-device-sound/src/device.rs
@@ -823,11 +823,12 @@ impl Drop for ControlMessage {
 
 #[cfg(test)]
 mod tests {
-    use super::*;
-    use crate::BackendType;
     use tempfile::tempdir;
     use vm_memory::GuestAddress;
 
+    use super::*;
+    use crate::BackendType;
+
     const SOCKET_PATH: &str = "vsound.socket";
 
     #[test]
diff --git a/staging/vhost-device-sound/src/lib.rs b/staging/vhost-device-sound/src/lib.rs
index a7b0f5a..a395dab 100644
--- a/staging/vhost-device-sound/src/lib.rs
+++ b/staging/vhost-device-sound/src/lib.rs
@@ -321,11 +321,13 @@ pub fn start_backend_server(config: SoundConfig) {
 
 #[cfg(test)]
 mod tests {
-    use super::*;
-    use crate::ControlMessageKind;
     use std::sync::{Arc, RwLock};
+
     use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
 
+    use super::*;
+    use crate::ControlMessageKind;
+
     #[test]
     fn test_sound_server() {
         const SOCKET_PATH: &str = "vsound.socket";
@@ -344,7 +346,8 @@ mod tests {
 
         let vring_workers = daemon.get_epoll_handlers();
 
-        // VhostUserSoundBackend support a single thread that handles the TX and RX queues
+        // VhostUserSoundBackend support a single thread that handles the TX and RX
+        // queues
         assert_eq!(backend.read().unwrap().threads.len(), 1);
 
         assert_eq!(vring_workers.len(), backend.read().unwrap().threads.len());
diff --git a/staging/vhost-device-sound/src/stream.rs b/staging/vhost-device-sound/src/stream.rs
index c57e6c9..be157a4 100644
--- a/staging/vhost-device-sound/src/stream.rs
+++ b/staging/vhost-device-sound/src/stream.rs
@@ -285,8 +285,6 @@ impl Drop for Buffer {
 
 #[cfg(test)]
 mod tests {
-    use super::*;
-    use crate::SoundDescriptorChain;
     use vhost_user_backend::{VringRwLock, VringT};
     use virtio_bindings::bindings::virtio_ring::{VRING_DESC_F_NEXT, VRING_DESC_F_WRITE};
     use virtio_queue::{mock::MockSplitQueue, Descriptor, Queue, QueueOwnedT};
@@ -294,6 +292,9 @@ mod tests {
         Address, ByteValued, GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap,
     };
 
+    use super::*;
+    use crate::SoundDescriptorChain;
+
     // Prepares a single chain of descriptors for request queue
     fn prepare_desc_chain<R: ByteValued>(
         start_addr: GuestAddress,
```

</details>

File: staging/vhost-device-sound/Cargo.toml [hunk 29]
```diff
@@ -10,6 +10,7 @@ license = "Apache-2.0 OR BSD-3-Clause"
 edition = "2018"
 
 [features]
+xen = ["vm-memory/xen", "vhost/xen", "vhost-user-backend/xen"]
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/Cargo.toml:13]: I'd put this change in a separate commit

File: staging/vhost-device-sound/Cargo.toml [hunk 30]
```diff
@@ -21,13 +22,15 @@ env_logger = "0.10"
 log = "0.4"
 pw = { package = "pipewire", git = "https://gitlab.freedesktop.org/pipewire/pipewire-rs.git", rev = "5fe090b3ac8f6fed756c4871ac18f26edda3ac89", optional = true }
 thiserror = "1.0"
-vhost = { version = "0.6", features = ["vhost-user-slave"] }
-vhost-user-backend = "0.8"
+vhost = { version = "0.8", features = ["vhost-user-slave"] }
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/Cargo.toml:25]: Also these updates

File: staging/vhost-device-sound/Cargo.toml [hunk 31]
```diff
@@ -21,13 +22,15 @@ env_logger = "0.10"
...
-virtio-queue = "0.7"
-vm-memory = "0.10"
+virtio-queue = "0.9"
+vm-memory = "0.12"
 vmm-sys-util = "0.11"
 
 [dev-dependencies]
-serial_test = "1.0"
 rstest = "0.18.2"
+tempfile = "3.5"
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/Cargo.toml:34]: These are fine to be in the same commit where you add the tests.

File: staging/vhost-device-sound/src/audio_backends/pipewire.rs [hunk 32]
```diff
@@ -161,6 +161,7 @@ impl AudioBackend for PwBackend {
 
     fn prepare(&self, stream_id: u32) -> Result<()> {
         debug!("pipewire prepare");
+        debug!("stream_id {}", stream_id);
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/src/audio_backends/pipewire.rs:164]: Unrelated?

File: staging/vhost-device-sound/src/device.rs [hunk 33]
```diff
@@ -678,7 +684,7 @@ impl VhostUserSoundBackend {
     }
 }
 
-impl VhostUserBackend<VringRwLock, ()> for VhostUserSoundBackend {
+impl VhostUserBackendMut<VringRwLock, ()> for VhostUserSoundBackend {
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/src/device.rs:687]: Why this change? I'd put in a separate commit with a description.

File: staging/vhost-device-sound/src/lib.rs [hunk 34]
```diff
@@ -285,8 +287,9 @@ impl Drop for IOMessage {
 pub fn start_backend_server(config: SoundConfig) {
     log::trace!("Using config {:?}", &config);
     let listener = Listener::new(config.get_socket_path(), true).unwrap();
-    let backend = Arc::new(VhostUserSoundBackend::new(config).unwrap());
-
+    let backend = Arc::new(RwLock::new(
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/src/lib.rs:290]: Also this change in a separate commit with a description of the reason.

**Review by stefano-garzarella — CHANGES_REQUESTED** [review]
Thanks for working on the test! Very much appreciated!

The tests LGTM, but I suggest splitting the code changes in multiple commits with a good explanation of the changes and the reason.

File: staging/vhost-device-sound/src/device.rs (hunk 33 above)
Comment (stefano-garzarella) [reply staging/vhost-device-sound/src/device.rs:687]: > Looks like it's from the dependency updates.
> 
> @dorindabassey my suggestion is first update the dependencies in a standalone commit, and add the test changes in a second one.

Yes, please ;-)

Comment (dorindabassey) [reply staging/vhost-device-sound/src/device.rs:687]: looks like this change is not necessary. I changed it because of this [test](https://github.com/rust-vmm/vhost-device/pull/489/files/d7179cc9091887f20caebf1c810332a87769108e#diff-ab6dc4ed31cc538191be45fbe8e1bce7df597f769bea01d2390a96faeffa0700R351)
afterall I can do `assert_eq!(backend.threads.len(), 1);`

I will split the commits

File: staging/vhost-device-sound/Cargo.toml (hunk 29 above)
Comment (dorindabassey) [reply staging/vhost-device-sound/Cargo.toml:13]: done

File: staging/vhost-device-sound/src/device.rs [hunk 35]
```diff
@@ -128,10 +128,13 @@ impl VhostUserSoundThread {
         vring: &VringRwLock,
         audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
     ) -> IoResult<bool> {
+        let Some(ref atomic_mem) = self.mem else {
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/src/device.rs:131]: I'd move the changes to the code related to the new ` NoMemoryConfigured` in a separate commit.

**Review by stefano-garzarella — COMMENTED** [review]
I'd also suggest to use just `sound:`, `sound\test:`, etc. as commit prefix, just to have a shorter title.

Final changes LGTM, I'd suggest just another small commit split and we should be fine IMHO.

**Review by stsquad — DISMISSED** [review]

File: staging/vhost-device-sound/src/device.rs (hunk 35 above)
Comment (dorindabassey) [reply staging/vhost-device-sound/src/device.rs:131]: done

**Review by stefano-garzarella — APPROVED** [review]
LGTM! Thanks!

**Review by vireshk — APPROVED** [review]
